libbpf-tools/ksnoop.bpf.c

```diff
@@ -0,0 +1,454 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2021, Oracle and/or its affiliates. */
+
+#include "vmlinux.h"
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+
+#include "ksnoop.h"
+
+/* For kretprobes, the instruction pointer in the struct pt_regs context
+ * is the kretprobe_trampoline.  We derive the instruction pointer
+ * by pushing it onto a function stack on entry and popping it on return.
+ *
+ * We could use bpf_get_func_ip(), but "stack mode" - where we
+ * specify functions "a", "b and "c" and only want to see a trace if "a"
+ * calls "b" and "b" calls "c" - utilizes this stack to determine if trace
+ * data should be collected.
+ */
+#define FUNC_MAX_STACK_DEPTH	16
+
+#ifndef ENOSPC
+#define ENOSPC			28
+#endif
+
+struct func_stack {
+	__u64 task;
+	__u64 ips[FUNC_MAX_STACK_DEPTH];
+	__u8 stack_depth;
+};
+
+/* function call stack hashed on a per-task key */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 512);
+	__type(key, __u64);
+	__type(value, struct func_stack);
+} ksnoop_func_stack SEC(".maps");
```

Comment:
Yep, the map is populated with one entry per function specified on the commandline, and there is no need to delete those entries - they live for the lifetime of the program to record tracing information for the function traced. max_entries should probably be lower now I look at it;  we only allow MAX_FUNC_TRACES (currently 8) different functions to be traced. I think we can probably bump MAX_FUNC_TRACES up to 64 and use it explicitly in the map max_entries to reflect the limitation. We probably don't want BPF_F_NO_PREALLOC since the map size is small I think?

Reply:
My concern here is, if we have 1000(> 512) process trigger the same kernel function, we may suffer data loss.

Reply:
hmm, for now I've made a change to make size 2048; since each map entry is ~150 bytes we can afford to do this I think.